libcudacxx/include/cuda/std/__ranges/single_view.h

```diff
@@ -0,0 +1,153 @@
+
+public:
+#if !defined(_CCCL_NO_CONCEPTS)
+  _CCCL_HIDE_FROM_ABI single_view()
+    requires default_initializable<_Tp>
+  = default;
+#else // ^^^ !_CCCL_NO_CONCEPTS ^^^ / vvv _CCCL_NO_CONCEPTS vvv
+  _CCCL_TEMPLATE(class _Tp2 = _Tp)
+  _CCCL_REQUIRES(default_initializable<_Tp2>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr single_view() noexcept(is_nothrow_default_constructible_v<_Tp>)
+      : view_interface<single_view<_Tp>>()
+      , __value_(){};
+#endif // _CCCL_NO_CONCEPTS
+
+  _CCCL_TEMPLATE(class _Tp2 = _Tp) // avoids circular concept definitions with copy_constructible
+  _CCCL_REQUIRES((!is_same_v<remove_cvref_t<_Tp2>, single_view>) _CCCL_AND copy_constructible<_Tp2>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr explicit single_view(const _Tp2& __t) noexcept(
+    is_nothrow_copy_constructible_v<_Tp2>)
+      : view_interface<single_view<_Tp2>>()
+      , __value_(in_place, __t)
+  {}
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr explicit single_view(_Tp&& __t) noexcept(is_nothrow_move_constructible_v<_Tp>)
+      : view_interface<single_view<_Tp>>()
+      , __value_(in_place, _CUDA_VSTD::move(__t))
+  {}
+
+  _CCCL_TEMPLATE(class... _Args)
+  _CCCL_REQUIRES(constructible_from<_Tp, _Args...>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr explicit single_view(in_place_t, _Args&&... __args) noexcept(
+    is_nothrow_constructible_v<_Tp, _Args...>)
+      : view_interface<single_view<_Tp>>()
+      , __value_{in_place, _CUDA_VSTD::forward<_Args>(__args)...}
+  {}
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp* begin() noexcept
+  {
+    return data();
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr const _Tp* begin() const noexcept
+  {
+    return data();
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp* end() noexcept
+  {
+    return data() + 1;
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr const _Tp* end() const noexcept
+  {
+    return data() + 1;
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI static constexpr size_t size() noexcept
+  {
+    return 1;
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp* data() noexcept
+  {
+    return __value_.operator->();
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr const _Tp* data() const noexcept
+  {
+    return __value_.operator->();
+  }
+};
+
+template <class _Tp>
+_CCCL_HOST_DEVICE single_view(_Tp) -> single_view<_Tp>;
+
+_LIBCUDACXX_END_NAMESPACE_RANGES
+
+_LIBCUDACXX_BEGIN_NAMESPACE_VIEWS
+_LIBCUDACXX_BEGIN_NAMESPACE_CPO(__single_view)
+
+template <class _Range>
+_CCCL_CONCEPT __can_single_view = _CCCL_REQUIRES_EXPR((_Range))(typename(single_view<decay_t<_Range>>));
+
+struct __fn : __range_adaptor_closure<__fn>
+{
+  _CCCL_TEMPLATE(class _Range)
+  _CCCL_REQUIRES(__can_single_view<_Range>) // MSVC breaks without it
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr auto operator()(_Range&& __range) const
+    noexcept(noexcept(single_view<decay_t<_Range>>(_CUDA_VSTD::forward<_Range>(__range))))
```

Comment:
is this simpler?
```c++
noexcept(_CUDA_VSTD::is_nothrow_constructible_v<_CUDA_VSTD::decay_t<_Range>, _Range>)
```

not really. but this comes up so often that i usually:

```c++
template <class... Ts>
concept __nothrow_decay_copyable =
  (is_nothrow_constructible_v<decay_t<Ts>, Ts> &&...);
```

very handy, that.